Add a new `key_estimator` module in Lua

elmaxxo opened the pull request:
This pull request implements a new module `key_estimator`, that uses the HyperLogLog algorithm for key estimating.

elmaxxo committed:
salad: implement the HyperLogLog algorithm

Implement the HyperLogLog algorithm based on the article
"HyperLogLog in Practice: Algorithmic Engineering of a
State of The Art Cardinality Estimation Algorithm".
The algorithm allows to estimate the cardinaliry of a multiset using
almost no memory overhead.

The realisation supports:
6-bit registers;
Dense representation;
Sparse representation;
Bias correction based on interpolated curves;

NO_CHANGELOG=internal
NO_DOC=internal

There are some benchmark results obtained in my configuration
(i5-10400 12 X 2904 MHz CPU):

Dense/Adding: ~14M-15M operations per second;
Dense/Estimating*: ~83-270K operations per second;
Sparse/Adding*: ~7.6M-9.8M operations per second;
Sparse/Estimating*: ~383-150K operations per second;

Dense/Estimating*:
estimating has a complexity of o(2^precision), so the best performance
is achieved with the smallest precision and the worst performance is
achieved with the highest precision.

Sparse/Adding*:
Average adding complexity is o(log(n)), where n is the number of added
pairs. The results are obtained by filling an empty estimator close to
its maximal size, which depends on the precision. The best results
correspond to the smaller precision, and the worst results correspond to
the highest precision.

Sparse/Estimating*:
Estimating has a complexity of o(n*log(n)), where n is the number of
pairs added to the estimator. The results are obtained by estimating
the cardinality close to the maximal size, which depends on the
precision. The best results correspond to the smaller precision, and
the worst results correspond to the highest precision.

elmaxxo committed:
lua: add key_estimator module

@TarantoolBot document
Title: key_estimator module API

key_estimator module can be used for estimating cardinality of
multiset of keys using almost no memory overhead.
The module uses the HyperLogLog algorithm, so the esimation error and
the memory usage are determined by the folowing formulas:

standard_error = 1.04 / sqrt(2^prec)
memory_usage = 2^precision * 6 / 8 bytes

The algorithm supports 2 representaions: sparse and dense.
The dense representation has constant memory usage and the error
determined by the formulas above.
Sparse representation memory usage is linearly proroptional to the
number of added distinct keys and provides the best accuracy.
Sparse representation can be used only for small cardinalities
(< 2^precision / 6). If the memory usage reaches the usage of the dense
representation the representation automatically switches to the
dense with precision defined when creating the estimator.

Module constatns:

* key_estimator:MAX_PRECISION - maximum value of the accuracy parameter.
* key_estimator:MIN_PRECISION - minimum value of the accuracy parameter.
* key_estimator:SPARSE_RPECISION - the precision that is used in the
				   sparse representation.
* key_estimator:DENSE - specify dense representation for the estimator in
			key_estimator.new funciton.
* key_estimator:SPARSE - specify sparse representation for the estimator
			 in key_estimator.new funciton.

Module methods:

key_estimator.new(key_fmt[, prec, repr])
Create a key_estimator object.

Parameters:
* key_fmt - format of the key for added tuples.
* prec - precision parameter for the HyperLogLog algorithm, determines
	 the memory usage and the estimation error.
* repr - representation of the HyperLogLog algorithm.

Returns:
key_estimator_object object.

Errors:
Invalid precision value.

key_estimator_object methods:

key_estimator_object:add(tuple)
Add key of the tuple to the estimator.
The complexity depends on the representation.
For the dense representation the complexity is o(1).
For the sparse representation the average complexity is o(log(n)).
In most cases the complexity is o(1), but in the worst
case it is o(n*log(n)), where n is the number of added keys.

Parameters:
* tuple - lua table or box.tuple whose key is added to the estimator.

Returns:
No return value.

Errors:
Bad key format.

key_estimator_object:estimate()
Estimate cardinality of the set of added keys.
The complexity depends on the representation.
For the dense representation the complexity is o(2^precision) = o(1).
For the sparse representation the complexity is o(n*log(n)), where n is
the number of added keys.

Returns:
Estimtation.

Errors:
No errors

key_estimator_object:merge(other_key_estimator_object)
Add all keys from other_key_estimator_object to the caller object.
The estimators must have the same precision value and key format.
If the caller estimator has the sparse representation it will be
changed to dense.
The complexity is linear to the number of distinct keys added to both
estimators.

Parameters:
* other_key_estimator_object - key estimator object.

Returns:
No return value.

Errors:
Different precision values. Different key formats.

CuriousGeorgiy commented on src/box/lua/key_estimator.c:
```diff
@@ -0,0 +1,216 @@
+
+	int rc = key_estimator_merge(dst, src);
+	if (rc != 0)
+		return luaT_error(L);
```
I think there should be some kind of diagnostic here.

CuriousGeorgiy commented on src/box/lua/key_estimator.c:
```diff
@@ -0,0 +1,216 @@
+	int rc = key_estimator_add(estimator, tuple);
+	tuple_unref(tuple);
+	if (rc != 0)
+		return luaT_error(L);
```
I think there should be some kind of diagnostic here too.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+    local min_prec = key_estimator.MIN_PRECISION
+    local max_prec = key_estimator.MAX_PRECISION
+    for prec = min_prec, max_prec, 1 do
+        local est = key_estimator.new(fmt, prec)
```
Please test both representations. You can do this using [test parametrization](https://github.com/tarantool/luatest#parametrization).

KaitmazianMO replied:
I made 2 tests for both representations.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+        local est = key_estimator.new(fmt, prec)
+        local err = hll_error(prec)
+
+        local card = 10 * 2 ^ prec
```
Please supply this formula with a comment.

KaitmazianMO replied:
Done.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+        end
+
+        --- 99.7% to lie in expected interval
+        local margin = card * err * 3
```
Please supply this formula with a comment.

KaitmazianMO replied:
Done.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+        local est2 = key_estimator.new(fmt, prec, key_estimator.DENSE)
+        local err = hll_error(prec)
+
+        local card = 5 * 2 ^ prec
```
Please supply this formula with a comment.

KaitmazianMO replied:
Done.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+            else
+                est1:add({tostring(i)})
+                est2:add({tostring(i)})
+            end
```
I don't quite understand this logic.

KaitmazianMO replied:
For this test, I divided the set range [0, map] between 2 estimators. Some items belong to only one estimator, and some belong to both of them. For the completeness of the test, we need to check that adding common elements does not change the estimation, unlike adding new ones. This is achieved by merging estimators that have some  common items and some unique items.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+
+        est1:merge(est2)
+        --- 99.7% to lie in expected interval
+        local margin = card * err * 3
```
Please supply this formula with a comment.

KaitmazianMO replied:
Done.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+    local est1 = key_estimator.new(fmt, key_estimator.MIN_PRECISION)
+    local est2 = key_estimator.new(fmt, key_estimator.MAX_PRECISION)
+    --- different precision
+    t.assert_error(key_estimator.merge, est1, est2)
```
Please use `assert_error_msg_equals`.

KaitmazianMO replied:
Done.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+    for i = 1, table.getn(estimators), 1 do
+        for j = i + 1, table.getn(estimators), 1 do
+            --- different formats
+            t.assert_error(key_estimator.merge, estimators[i], estimators[j])
```
Please use `assert_error_msg_equals`.

KaitmazianMO replied:
Done.

CuriousGeorgiy commented on changelogs/unreleased/add-key-estimator-module.md:
```diff
@@ -0,0 +1,3 @@
+## featule/lua
```
```suggestion
## feature/lua
```

KaitmazianMO replied:
Thank you for noticing the typo.

KaitmazianMO commented on src/lua/utils.h:
```diff
@@ -258,6 +258,21 @@ void
+	lua_pushstring(L, varriable_name);
+	lua_pushnumber(L, varriable_value);
+	lua_settable(L, -3);
+}
```
How can I reuse macro from .c file?
I can't use this macro in `luaL_add_varriable` because  `lua_pushliteral` can't accept a string as a pointer.

CuriousGeorgiy replied:
I mean you can move this macro to `src/lua/utils.h` and then use it in your patch: I didn't find any usages of non-literal strings in `luaL_add_variable`.

KaitmazianMO replied:
Moved this method to lua/utils.h.

KaitmazianMO commented on src/box/lua/key_estimator.c:
```diff
@@ -0,0 +1,216 @@
+	 * declared. The standard error is less than 1%.
+	 */
+	DEFAULT_PRECISION = 14,
+};
```
Done.

KaitmazianMO commented on src/box/key_estimator.c:
```diff
@@ -0,0 +1,115 @@
+	  * In addition, using PMurHash32 helps to randomize hashes for integer
+	  * values whose hash is equal to the input integer value.
+	  */
+	uint64_t h1 = PMurHash32(0, &hash, sizeof(hash));
```
Done.

KaitmazianMO commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+local key_estimator = require('key_estimator')
+
+local function hll_error(prec)
+    return 1.04 / (2 ^ (prec / 2.0))
```
Done.

KaitmazianMO commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+
+g.test_common = function()
+    --- expect key format
+    t.assert_error(key_estimator.new)
```
Done.

KaitmazianMO commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+    local est = key_estimator.new(fmt)
+
+    --- invalid precision
+    t.assert_error(key_estimator.new, fmt, key_estimator.MAX_PRECISION + 1)
```
Done.

KaitmazianMO commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+    t.assert_equals(est:estimate(), 1)
+
+    --- wrong key format, string is expected
+    t.assert_error(key_estimator.add, est, {0})
```
Done.

KaitmazianMO commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,110 @@
+    t.assert_error(key_estimator.add, est, {0})
+end
+
+g.test_error = function()
```
I changed it to *estimation_error*.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,187 @@
+        local std_err = hll_std_error(prec)
+
+        --- cardinality for wich the HyperLogLog algorithm will be used.
+        local card = 5 * hll_n_counters(prec)
```
No, by explaining the formula I mean, where does `5` come from?

KaitmazianMO replied:
Added a more detailed description.

CuriousGeorgiy commented on test/app-luatest/key_estimator_test.lua:
```diff
@@ -0,0 +1,187 @@
+        local est = key_estimator.new(fmt, prec)
+        local std_err = hll_std_error(prec)
+
+        --- cardinality for wich the HyperLogLog algorithm will be used.
```
No, by explaining the formula I mean, where does `5` come from?

KaitmazianMO replied:
Added a more detailed description.

elmaxxo committed:
refactoring: move PUSHTABLE macro to lua/utils.h

Add a new luaL_tablepush macro to lua/utils.h and replace all PUSHTABLE
entries in fio.c by the new macro.

NO_DOC=refactoring
NO_TEST=refactoring
NO_CHANGELOG=refactoring

Totktonada commented on src/box/lua/init.c:
```diff
@@ -534,6 +535,8 @@ box_lua_init(struct lua_State *L)
 	lua_pop(L, 1);
 	luaopen_merger(L);
 	lua_pop(L, 1);
+	luaopen_key_estimator(L);
+	lua_pop(L, 1);
```
Do you consider other possible names? It is multiset, right? Why not `key_multiset` or `keyring`? (But it is ofter used in the security context.) Or just `keyset` for simplicity. How about `multikey`? (Don't know, whether it has any relation to tarantool's multikey indexes, but if it does, it would be perfect.)

I would also engage @locker into the discussion. He ofter has good ideas regarding APIs.

KaitmazianMO replied:
This is a HyperLogLog algorithm, so I think such names can be confusing.

Totktonada replied:
I misread the API: now I got that we add tuples into the multiset, not key_def objects. And it is not a container, it doesn't store the tuples.

Anyway, 'estimator' sounds too common here. What does it estimates? It is unclear from the name.

How about `key_def.hll` submodule?

KaitmazianMO replied:
I don't want to call it `hll` because the Hyperloglog algorithm works with hashes, but the module works with tuples, so in my opinion the name should explain what it does in general, for example, *estimator*. The module estimates keys, so I guess it should be mentioned in its name too.

Could you explain why do want to make it a submosule of `key_def`?

locker replied:
I don't like the names, either. TBH I don't understand why anyone would need this thing in Lua. (In C this might be helpful though - e.g. to estimate the number of unique elements in two merged Vinyl run files.) I think we should run this by the PM team.

Anyway, I would implement this module independently of box, tuples, or `key_def`. I think it should be named `hll` or `hyperloglog`, should be enabled with `require`, and should take hashes, like in the C API. To add tuples in HLL, we should add a function to the `key_def` module that takes a tuple and computes its hash. This way, the API would have clear names and would be more flexible, because we could add any hashable objects to it.

KaitmazianMO replied:
I like your suggestions. Should I start implementing them or do I need to wait for a response from the PM team?

If the first, what do you think about the design of `key_def:hash()`? If we want to use this hash for the algorithm, we can't just use `tuple_hash()` because it returns 32-bit values. It also doesn't randomize the hash for integers, so the estimation for sets like {1, 2, 3} will be 1.
But at the same time I think that if `key_def:hash()` returns something other than the `tuple_hash()`, it might be confusing.

locker replied:
PM @unera does not approve adding HyperLogLog to Tarantool Lua API.

CuriousGeorgiy commented:
Filed FP linter errors: tarantool/checkpatch#39 and tarantool/checkpatch#40.

locker commented on src/lua/utils.h:
```diff
@@ -258,6 +258,19 @@ void
+ * @param method method that pushes the value on the Lua stack.
+ * @param value value.
+ */
+#define luaL_tablepush(name, method, value) do	{	\
```
I don't like this helper, because it doesn't work if:
 - Lua state has other name than `L`. E.g. we use `tarantool_L` sometimes.
 - It doesn't work if the table isn't on top of the Lua stack.
 - It doesn't work if the method takes more than one argument.

Since it's a trivial macro, I think that you should simply redefine it in your source file instead.

locker commented on src/box/key_estimator.c:
```diff
@@ -0,0 +1,121 @@
+uint64_t
+key_estimator_estimate(struct key_estimator *estimator)
+{
+	return hll_estimate(estimator->hll);
```
Better name `hll_count_distinct` to emphasize that it estimates the count of distinct elements. BTW the term "cardinality" is confusing, because it means different things in different sources:

Quoting [wiki](https://en.wikipedia.org/wiki/Multiset):

> The [cardinality](https://en.wikipedia.org/wiki/Cardinality) of a multiset is the sum of the multiplicities of all its elements. For example, in the multiset {a, a, b, b, b, c} the multiplicities of the members a, b, and c are respectively 2, 3, and 1, and therefore the cardinality of this multiset is 6.

Mons commented:
Why is this not a separate (standalone) lua module?

CuriousGeorgiy commented:
@Mons because it requires the HyperLogLog data structure from salad.
